Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,194 @@
+public final class CacheAdvance<T: Codable> {
+
+    // MARK: Initialization
+
+    /// Creates a CacheAdvance.
+    /// - Parameters:
+    ///   - file: The URL for the on-disk cache to live.
+    ///   - maximumBytes: The maximum size of the cache, in bytes.
+    ///   - shouldRoll: When `true`, new logs will overwrite the oldest logs when the cache runs out of space.
+    public init(
+        file: URL,
+        maximumBytes: UInt32,
+        shouldRoll: Bool)
+        throws
+    {
+        self.file = file
+        self.maximumBytes = maximumBytes
+        self.shouldRoll = shouldRoll
+
+        writer = try FileHandle(forWritingTo: file)
+        reader = try FileHandle(forReadingFrom: file)
+
+        lengthOfMessageSuffix = {
+            if shouldRoll {
+                /// We store both the `endOfNewestMessageMarker` and `offsetInFileOfOldestMessage` after each message.
+                return 2 * Data.messageSpanLength
+            } else {
+                /// We store a `endOfNewestMessageMarker` after each message.
+                return Data.messageSpanLength
+            }
+        }()
+    }
+
+    deinit {
+        try? writer.close()
+    }
+
+    // MARK: Public
+
+    public let file: URL
+
+    /// Appends a message to the cache.
+    /// - Parameter message: A message to write to disk.
+    /// - Returns: Whether there was room in the cache to append the message. Always `true` for caches that roll.
+    @discardableResult
+    public func append(message: T) throws -> Bool {
+        try setUpFileHandlesIfNecessary()
+
+        let encodedMessage = EncodableMessage(message: message, encoder: encoder)
+        let messageData = try encodedMessage.encodedData()
+        let cacheHasSpaceForNewMessage = writer.offsetInFile + UInt64(messageData.count + lengthOfMessageSuffix) <= UInt64(maximumBytes)
+        if cacheHasSpaceForNewMessage {
+            try write(messageData: messageData)
+            return true
+
+        } else if shouldRoll {
+            // Trim the file to the current position.
+            try writer.truncate(atOffset: writer.offsetInFile)
+
+            // Set the offset back to the beginning of the file.
+            try writer.seek(toOffset: 0)
+
+            // We know the oldest message is at the beginning of the file, since we just tossed out the rest of the file.
+            try reader.seek(toOffset: 0)
+            // We know we're about to overwrite the oldest message, so advance the reader to the next message.
+            try reader.seekToNextMessage(shouldSeekToOldestMessageIfFound: false)
+
+            // Start writing from the beginning of the file.
+            try write(messageData: messageData)
+            return true
+
+        } else {
+            // We're out of room.
+            return false
+        }
+    }
+
+    /// Fetches all messages from the cache.
+    public func cachedMessages() throws -> [T] {
```

Comment:
IMO returning an Iterable type makes more sense here, since the store could be enormous. Feels like it could be too expensive / unnecessary to actually read back everything at once. Not sure what the state of this is in Swift these days but `-> CacheIterator` where `CacheIterator : IteratorProtocol` that walks over all cached entries would be nice, and if you want them all, you could just `collect()` them into an `[T]`.

Reply:
Makes sense. You could also just make an interface that consumes a single item, and have another one that consumes all of them via the new interface.

Reply:
Indeed! These improvements are in the We’ll Get There™ camp 🙂